pkg/visitors/assemble.go

```diff
@@ -71,6 +72,86 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
 		}
 
 		fileOffset := f.DataOffset
+
+		var pad *uefi.File
+		var shrink uint64
+		for _, file := range f.Files {
+			fileBuf := file.Buf()
+			fileLen := uint64(len(fileBuf))
+			if fileLen == 0 {
+				log.Fatal("Unexpected empty file buffer (at least a header was expected): %v", file.Header.GUID)
+			}
+			if file.Header.GUID == *uefi.FFGUID && fileLen > 24 {
+				// found non-empty PAD to check for free space.
+				pad = file
+			}
+
+			// Pad to the 8 byte alignments.
+			alignedOffset := uefi.Align8(fileOffset)
+			// Read out the file alignment requirements
+			if alignBase := file.Header.Attributes.GetAlignment(); alignBase != 1 {
+				hl := file.HeaderLen()
+				// We need to align the data, not the header. This is so terrible.
+				fileDataOffset := uefi.Align(alignedOffset+hl, alignBase)
+				// Calculate the starting offset of the file
+				newOffset := fileDataOffset - hl
+				if gap := (newOffset - alignedOffset); gap >= 8 && gap < uefi.FileHeaderMinLength {
+					// We need to re align to the next boundary cause we can't put a pad file in here.
+					// Who thought this was a good idea?
+					fileDataOffset = uefi.Align(fileDataOffset+1, alignBase)
+					newOffset = fileDataOffset - hl
+				}
+				if newOffset != alignedOffset {
+					// Add a pad file starting from alignedOffset to newOffset
+					alignedOffset = newOffset
+				}
+			}
+			fileOffset = alignedOffset + fileLen
+		}
+
+		// Check if we're out of space.
+		if f.Length < fileOffset && !f.Resizable {
+			// do we have a non-empty pad file?
+			if pad != nil {
+				shrink = fileOffset - f.Length
```

Comment:
This will break many bioses. Non empty pad files are often used to align other modules.

Reply:
The intent is not changing the ending address of the pad file since EFI driver is added so the alignment will be unchanged.